src/docfx/lib/msgraph/MicrosoftGraphCache.cs

```diff
@@ -43,57 +41,52 @@ public MicrosoftGraphCache(Config config)
             }
         }
 
-        public Task<(Error error, MicrosoftAlias msAlias)> GetMicrosoftAliasAsync(string alias)
+        public async Task<(Error error, MicrosoftAlias msAlias)> GetMicrosoftAliasAsync(string alias)
         {
             if (string.IsNullOrEmpty(alias))
             {
-                return new Task<(Error error, MicrosoftAlias msAlias)>(null, null);
+                return default;
             }
 
-            return Synchronized(GetAsyncCore);
+            if (_aliases.TryGetValue(alias, out var msAlias))
+            {
+                return (null, msAlias);
+            }
 
-            async Task<(Error error, MicrosoftAlias msAlias)> GetAsyncCore()
+            if (_tempInvalidAliases.Contains(alias) || _microsoftGraphAccessor == null)
```

Comment:
ConcurrentBag

Reply:
@Osmond, for each over concurrentbag creates a snapshot of all items, Contains method is using LINQ, everytime contains is called, the collection is duplocated. We have ConcurrentHashset for this purpose

Reply:
I have offline talked with @LowEntropyBody , he should use ConcurrentHashset